### PR TITLE
Enable http based plugin extensions

### PR DESCRIPTION
**Pull request by aquamatthias**
By defining an abstract http handler trait, as well as a PluginsResource to handle plugin requests.
Idea:
-GET /v2/plugins will list all available HttpHandler implementations
-GET/HEAD/PUT/POST/DELETE /v2/plugins/<context_path>/xxx/yyy will be handled by the plugin  that handles the defined context path (defined by the plugin). The path /xxx/yyy has to be handled by the plugin.

I will create an example in the marathon-example-plugins project.

@orlandohohmeier @aldipower: comments welcome

### CODE REVIEW

**Comment by aquamatthias** [comment]
Example plugin can be found here: mesosphere/marathon-example-plugins#8

**Comment by gkleiman** [comment]
I like what I see =)

**Comment by orlandohohmeier** [comment]
Yeah! This is very near to what I had in mind thinking about the Marathon UI Plugin integration. Thanks a bunch :bouquet: 

I would like to propose a slightly adjusted plugin configuration as well as path mapping and loading process,  after playing with your WIP implementation.

## Context Path

Consider having two plugins serving files, if both of them use the same `contextPath` (e.g. `/static`) you might get unwanted results. This could easily avoided if we define (unique) plugin ids and use those as context path.

This is how `plugin-conf.json` with ids could like:
 
```
{
  "experiment": {
    "plugin": "mesosphere.marathon.plugin.http.HttpRequestHandler",
    "implementation": "mesosphere.marathon.example.plugin.experiment.WebJarHandler"
  },
  "http": {
    "plugin": "mesosphere.marathon.plugin.http.HttpRequestHandler",
    "implementation": "mesosphere.marathon.example.plugin.http.WebJarHandler"
  },
  "example-authorizer": {
    "plugin": "mesosphere.marathon.plugin.auth.Authorizer",
    "implementation": "my.company.ExampleAuthorizer"
  },
  "example-authenticator": {
    "plugin": "mesosphere.marathon.plugin.auth.Authenticator",
    "implementation": "my.company.ExampleAuthenticator",
    "configuration": {
      "users": [
        {
          "user": "ernie",
          "password": "ernie",
          "permissions": [
            {
              "allowed": "create",
              "on": "/dev/"
            }
          ]
        }
      ]
    }
  }
}
```

*BWT: I was wondering if we really need all the plugin/implementation stuff...*

## Isolation

At the moment I have a one plugin serving just some files (static) and another one serving a gui plugin, but due to the lack of isolation both serve the same files as they use the same resource path (`/META-INF/resources/webjars/static/`). 

Plugin endpoint (`/v2/plugins`) response:              

```
[
    {
        "contextPath": "gui", 
        "implementation": "mesosphere.marathon.example.plugin.gui.WebJarHandler"
    }, 
    {
        "contextPath": "static", 
        "implementation": "mesosphere.marathon.example.plugin.http.WebJarHandler"
    }
]
```

Static Plugin (`/v2/plugins/static/package.json`) response:
```
{	
	"description": "Example Plugin", 
    "main": "main.js" 
}
```

Gui Plugin (`/v2/plugins/gui/package.json`) response:
```
{	
	"description": "Example Plugin", 
    "main": "main.js" 
}
```


Is there a way to restrict/isolate them? It would also be nice if only plugins (jars) that are actually defined within the `plugin-conf.json` are loaded as the active plugins even serve files from inactive plugins. 

--

What do you think?

**Comment by aquamatthias** [comment]
@orlandohohmeier: Cool! And thanks for the feedback.

Let me elaborate a little bit on the plugin mechanism:
- Plugins are (java/scala) classes identified by name. 
- There can be only one class with the same name.
- It would be possible to have multiple instances of the same class.
Having two plugins defining the same class will not work. There can be multiple plugins that implement the same plugin interface, but the class has to be different. 

The contextPath is defined in the plugin interface itself: [WebJarHandler](https://github.com/mesosphere/marathon-example-plugins/blob/mv/http_example/http/src/main/scala/mesosphere/marathon/example/plugin/http/WebJarHandler.scala#L16)
So your example of having:
-`mesosphere.marathon.example.plugin.gui.WebJarHandler` serving `/v2/plugins/static`
-`mesosphere.marathon.example.plugin.http.WebJarHandler` serving `/v2/plugins/gui` 
would work perfectly well. It would be also very easy to define the contextPath in the configuration.

Example:
```json
{
  "plugins": [
    {
      "plugin": "mesosphere.marathon.plugin.http.HttpRequestHandler",
      "implementation": "mesosphere.marathon.example.plugin.http.WebJarHandler",
      "configuration": {
        "contextPath": "gui"
      }
    }
  ]
}
```

I like your proposal of defining the list of plugins as objects with unique identifier instead of an array.
Need to think about it.

**Comment by orlandohohmeier** [comment]
@aquamatthias The current implementation does not serve javascript files. I always get `500 Internal Server Error` with the following JSON response body: `{ "message": null }`. Could you please have a look?

**Comment by aquamatthias** [comment]
@orlandohohmeier good catch. The problem was the media mime mapping - the nio logic does not work for zip file entries (I still do not understand why). I replaced that part - should work now.

**Comment by orlandohohmeier** [comment]
@aquamatthias Thanks for the fix! Works like a charm :100:

**Comment by aquamatthias** [comment]
@gkleiman any updates here?

File: plugin-interface/src/main/scala/mesosphere/marathon/plugin/http/HttpRequestHandlerBase.scala [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+import java.io.{ ByteArrayOutputStream, InputStream, OutputStream }
+import java.net.{ URL, URLConnection }
+import java.nio.file.{ Files, Paths }
+
+import scala.annotation.tailrec
+import scala.util.Try
+
+//scalastyle:off magic.number
+abstract class HttpRequestHandlerBase extends HttpRequestHandler {
```
Comment (gkleiman) [inline plugin-interface/src/main/scala/mesosphere/marathon/plugin/http/HttpRequestHandlerBase.scala]: Maybe add some comments describing the class?

File: plugin-interface/src/main/scala/mesosphere/marathon/plugin/http/HttpRequestHandlerBase.scala [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+  protected[this] def serveResource(path: String, response: HttpResponse): Unit = {
+    def transfer(in: InputStream, out: OutputStream) {
```
Comment (gkleiman) [inline plugin-interface/src/main/scala/mesosphere/marathon/plugin/http/HttpRequestHandlerBase.scala]: Have you thought about using Streams + transferTo? The code would look cleaner and I think it uses sendFile under-the-hood, so it would also perform better.

File: scalastyle-config.xml [hunk 3]
```diff
@@ -3,7 +3,7 @@
  <check level="error" class="org.scalastyle.file.FileTabChecker" enabled="true"/>
  <check level="error" class="org.scalastyle.file.FileLengthChecker" enabled="true">
   <parameters>
-   <parameter name="maxFileLength"><![CDATA[800]]></parameter>
```
Comment (gkleiman) [inline scalastyle-config.xml]: If we do this, then we can remove the `// scalastyle:off file.size.limit` line from Formats.scala

File: src/main/scala/mesosphere/marathon/api/v2/PluginsResource.scala [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+import mesosphere.marathon.MarathonConf
+import mesosphere.marathon.api.v2.json.Formats._
+import mesosphere.marathon.api.{ MarathonMediaType, RequestFacade, ResponseFacade, RestResource }
+import mesosphere.marathon.core.plugin.PluginsDescriptor
+import mesosphere.marathon.plugin.http.{ HttpRequest, HttpRequestHandler, HttpResponse }
+
+@Path("v2/plugins")
+class PluginsResource @Inject() (val config: MarathonConf,
+                                 requestHandler: Seq[HttpRequestHandler],
```
Comment (gkleiman) [inline src/main/scala/mesosphere/marathon/api/v2/PluginsResource.scala]: s/requestHandler/requestHandlers/

File: src/main/scala/mesosphere/marathon/api/v2/PluginsResource.scala [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+                                 descriptor: PluginsDescriptor) extends RestResource {
+
+  val handlerMap = descriptor.plugins
```
Comment (gkleiman) [inline src/main/scala/mesosphere/marathon/api/v2/PluginsResource.scala]: What about calling this pluginIdToHandler?

File: src/main/scala/mesosphere/marathon/api/v2/PluginsResource.scala [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    .filter(_.plugin == classOf[HttpRequestHandler].getName)
+    .flatMap { d => requestHandler.find(_.getClass.getName == d.implementation).map(d.id -> _) }
+    .toMap
+
+  @GET
+  @Produces(Array(MarathonMediaType.PREFERRED_APPLICATION_JSON))
+  def plugins(): Response = ok(jsonString(descriptor))
+
+  @GET
+  @Path("""{path}/{pluginPath:.+}""")
```
Comment (gkleiman) [inline src/main/scala/mesosphere/marathon/api/v2/PluginsResource.scala]: `@Path("""{pluginId}/{path:.+}""")` might be a bit clearer.

File: src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala [hunk 7]
```diff
@@ -782,3 +784,16 @@ trait V2Formats {
...
   )(Group(_, _, _, _, _), unlift(Group.unapply))
 }
+
+trait PluginFormats {
+
+  implicit lazy val pluginDefinitionFormat: Writes[PluginDefinition] = (
+    (__ \ "id").write[String] ~
+    (__ \ "plugin").write[String] ~
+    (__ \ "implementation").write[String] ~
+    (__ \ "tags").writeNullable[Set[String]] ~
```
Comment (gkleiman) [inline src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala]: what are the tags for?

File: src/main/scala/mesosphere/marathon/core/plugin/PluginDescriptor.scala [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+import play.api.libs.json.JsObject
+
+case class PluginDefinition(id: String,
+                            plugin: String,
+                            implementation: String,
+                            tags: Option[Set[String]],
+                            configuration: Option[JsObject],
+                            info: Option[JsObject])
+
+case class PluginsDescriptor(plugins: Seq[PluginDefinition])
```
Comment (gkleiman) [inline src/main/scala/mesosphere/marathon/core/plugin/PluginDescriptor.scala]: I find `PluginsDescriptor` and `PluginDefinition` maybe too similar. What do you think about renaming `PluginsDescriptor` to `PluginsConfiguration` or maybe just `Plugins`?

File: src/main/scala/mesosphere/marathon/core/plugin/impl/PluginManagerImpl.scala [hunk 9]
```diff
@@ -18,7 +23,7 @@ import scala.reflect.ClassTag
   * The plugin manager can load plugins from given urls.
   * @param urls the urls pointing to plugins.
   */
-private[plugin] class PluginManagerImpl(val descriptor: PluginsDescriptor, val urls: Seq[URL]) extends PluginManager {
+class PluginManagerImpl(val descriptor: PluginsDescriptor, val urls: Seq[URL]) extends PluginManager {
```
Comment (gkleiman) [inline src/main/scala/mesosphere/marathon/core/plugin/impl/PluginManagerImpl.scala]: Why did you change the visibility of this class? It doesn't seem like it is being used outside of the plugin package.

**Comment by gkleiman** [comment]
@aquamatthias sorry for the delay, made a couple of comments.

This also needs to be rebased and I think we also should update the plugins documentation and add something to the change log.

**Comment by orlandohohmeier** [comment]
I'd like to propose another plugin descriptor update which also supports plugins that not only use the `HttpRequestHandler` but also other plugin interfaces.

```
{
  "plugins": {
    "my-cool-plugin-id": {
      "implementations": {
            "mesosphere.marathon.plugin.http.HttpRequestHandler":
                "mesosphere.marathon.plugin.myCoolPlugin.HttpRequestHandler",
            "mesosphere.marathon.plugin.other.CoolPluginThing":
                "mesosphere.marathon.plugin.myCoolPlugin.other.CoolPluginThing"
        }
      "info": {
        "name": "My Cool Marathon Plugin",
        "modules": ["ui", "service"],
        "description": ""
      }
    }
  }
}
```

What do you think?

File: plugin-interface/src/main/scala/mesosphere/marathon/plugin/http/HttpRequestHandlerBase.scala (hunk 1 above)
Comment (aquamatthias) [inline plugin-interface/src/main/scala/mesosphere/marathon/plugin/http/HttpRequestHandlerBase.scala]: Done.

File: plugin-interface/src/main/scala/mesosphere/marathon/plugin/http/HttpRequestHandlerBase.scala (hunk 2 above)
Comment (aquamatthias) [inline plugin-interface/src/main/scala/mesosphere/marathon/plugin/http/HttpRequestHandlerBase.scala]: No. Can you make a simple example?

File: src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala (hunk 7 above)
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala]: Added a sentence in the documentation. Basically `tags` and `info` is there to add meta data to the plugin.

**Comment by aquamatthias** [comment]
@orlandohohmeier not sure what the goal of this proposal is: 
- one jar can provide multiple plugins. 
- there can be multiple implementations of the same plugin interface
Can you describe the problem/goal more specific?

Second: your proposal breaks identity: there is one id, but 2 plugins.

**Comment by orlandohohmeier** [comment]
@aquamatthias I know that it is possible to bundle several plugins in one jar. And it probably boils down to me still thinking about plugins as Marathon plugins and not `ServiceLoader` plugins. In my point of view using and configuring a plugin should be as easy as possible. I could think of cases in which one plugin or part of a plugin is useless without the other thus I think they should be defined and handle as one Marathon plugin. So yeah, it's one Marathon plugin (plugin id) and several  `ServiceLoader` plugins.

File: plugin-interface/src/main/scala/mesosphere/marathon/plugin/http/HttpRequestHandlerBase.scala (hunk 2 above)
Comment (gkleiman) [inline plugin-interface/src/main/scala/mesosphere/marathon/plugin/http/HttpRequestHandlerBase.scala]: Forget what I said about transferTo, that's only useful if you're writing directly to a socket (using a Channel).

But Marathon already depends on Guava, so we can get rid of this method and do:

```scala
val content = withResource(path) { url =>
  response.body(mediaMime(url), Resources.toByteArray(url))
  response.status(200)
}
content.getOrElse(response.status(404))
```

Comment (aquamatthias) [inline plugin-interface/src/main/scala/mesosphere/marathon/plugin/http/HttpRequestHandlerBase.scala]: Then guava becomes a dependency to all plugins.

Comment (aquamatthias) [inline plugin-interface/src/main/scala/mesosphere/marathon/plugin/http/HttpRequestHandlerBase.scala]: I tried `Files.copy` which looked promising. But it does not handle files from a zip file... :(

Comment (gkleiman) [inline plugin-interface/src/main/scala/mesosphere/marathon/plugin/http/HttpRequestHandlerBase.scala]: Since Marathon already depends on it, it would be a `provided` dependency, like logback.

Comment (aquamatthias) [inline plugin-interface/src/main/scala/mesosphere/marathon/plugin/http/HttpRequestHandlerBase.scala]: I agree - this makes it very easy. Thanks for the hint.

**Comment by aquamatthias** [comment]
@gkleiman Hope I addressed all comments. Squashed. Ready to merge :)

File: src/main/scala/mesosphere/marathon/core/plugin/PluginDescriptor.scala [hunk 10]
```diff
@@ -0,0 +1,12 @@
+package mesosphere.marathon.core.plugin
```
Comment (gkleiman) [inline src/main/scala/mesosphere/marathon/core/plugin/PluginDescriptor.scala]: Nit: The file is called `PluginDescriptor.scala`, but there's no `PluginDescriptor` class anymore.

File: src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala (hunk 7 above)
Comment (gkleiman) [inline src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala]: What's the goal of this meta data? It is not available to the plugins nor used by Marathon itself.

Is it to work around not being able to write comments in the configuration file? If that's the case, isn't `info` enough?

**Comment by gkleiman** [comment]
LGTM.

Just a comment about the renaming of PluginDescriptor to Plugins not being complete.

**Comment by aquamatthias** [comment]
@gkleiman Renamed the file and squashed again.

**Comment by aquamatthias** [comment]
@Renamed the parameter and all usages. Added RAML documentation.

**Comment by gkleiman** [comment]
:rocket:

Thanks! Waiting for the tests to pass before mergin'.
